velox/functions/lib/Re2Functions.cpp

```diff
@@ -666,6 +775,97 @@ FOLLY_ALWAYS_INLINE static bool isAsciiArg(
       rows);
 }
 
+// Match the input against a kRelaxedSubString pattern.
+// The algorithm is:
+// - Start by searching the first literal sub-pattern in the input.
+// - If we find the sub-pattern, we match the left & right part of the whole
+// pattern.
+// - If we failed matching any sub-pattern, we continue by searching the failed
+// literal sub-pattern.
+template <bool isAscii>
+bool matchRelaxedSubstringPattern(
+    const StringView& input,
+    const PatternMetadata& patternMetadata) {
+  if (input.size() < patternMetadata.length()) {
+    return false;
+  }
+
+  // Search the specified literal pattern in the input.
+  auto inputView = std::string_view(input);
+  auto patternStrHeader = patternMetadata.fixedPattern().c_str();
+  // The index in the input we are matching against.
+  auto currentMatchingOffsetInInput = 0;
+  auto searchLiteralPattern = [&](SubPatternMetadata& subPattern) -> int32_t {
+    auto index = inputView.find(
+        patternStrHeader + subPattern.start,
+        currentMatchingOffsetInInput,
+        subPattern.length);
+    if (index == std::string::npos) {
+      return -1;
+    }
+
+    return index;
+  };
+
+  // The index of the sub literal pattern we are matching.
+  auto currentPatternIndex =
+      (patternMetadata.subPatterns()[0].kind == kLiteralString ? 0 : 1);
+  auto currentSubPattern = patternMetadata.subPatterns()[currentPatternIndex];
+  auto end = input.size();
+  while (currentMatchingOffsetInInput < end) {
+    // Search current literal pattern to find out where to start matching
+    // the whole pattern.
+    currentMatchingOffsetInInput = searchLiteralPattern(currentSubPattern);
```

Comment:
Is this the only place where which advance currentMatchingOffsetInInput? How do we make sure we always advance and never get stuck?

Reply:
I'm wondering if we can advance currentMatchingOffsetInInput a bit more aggressively in some cases.

Let's say we look for sub-p N, find it at byte n, match to the right and find only partial match. I believe we can advance now to the start of the sub-pattern that failed to match and continue searching using that sub-pattern from there.

Reply:
Still wondering about this.